quark/core/shurikenapkinfo.py

```diff
@@ -0,0 +1,612 @@
+                )
+            )
+
+        return upperFuncs
+
+    @functools.lru_cache()
+    def lowerfunc(
+        self, method_object: MethodObject
+    ) -> list[Tuple[MethodObject, int]]:
+        """
+        Inherited from baseapkinfo.py.
+        Find the xrefs to the specified method.
+
+        :param method_object: a target method used to find what methods it
+        calls
+        :return: a set of tuples consisting of the called method and the
+        offset of the invocation
+        """
+        methodAnalysis = method_object.cache
+
+        lowerFuncs = []
+        for i in range(methodAnalysis.n_of_xrefto):
+            xref = methodAnalysis.xrefto[i]
+            lowerFuncs.append(
+                (
+                    self.__convert_to_method_object(xref.method.contents),
+                    xref.idx,
+                )
+            )
+
+        return lowerFuncs
+
+    def get_method_bytecode(
+        self, method_object: MethodObject
+    ) -> Generator[BytecodeObject, None, None]:
+        """
+        Inherited from baseapkinfo.py.
+        Return the bytecodes of the specified method.
+
+        :param method_object: a target method to get the corresponding
+        bytecodes
+        :yield: a generator of BytecodeObjects
+        """
+        disassembledMethod = self.__getDisassembledMethod(method_object)
+        if method_object.cache.external:
+            return
+        for i in range(disassembledMethod.n_of_instructions):
+            rawSmali = disassembledMethod.instructions[i].disassembly.decode(
+                errors="backslashreplace"
+            )
+            yield self.__parseSmali(rawSmali)
+
+    def __parseParameters(self, parameter: str) -> Union[int, float, str]:
+        if parameter[:2] == "0x":
+            try:
+                parameter = int(parameter, 16)
+                return parameter
+            except (TypeError, ValueError):
+                pass
+
+        try:
+            parameter = int(parameter, 10)
+            return parameter
+        except (TypeError, ValueError):
+            pass
+        try:
+            parameter = float(parameter)
+            return parameter
+        except (TypeError, ValueError):
+            pass
+
+        patternToIdentifyMethodCall = r"->\w+\("
+        if re.search(patternToIdentifyMethodCall, parameter):
+            parameter = self.__convertMethodCallFormat(parameter)
+
+        patternToIdentifyMemberField = r"->\w+(?!\() "
+        if re.search(patternToIdentifyMemberField, parameter):
+            parameter = self.__convertMemberFieldFormat(parameter)
+
+        return parameter
+
+    def __parseSmali(self, smali: str) -> BytecodeObject:
+        """
+        Parses the given smali code string into a BytecodeObject.
+
+        :param smali: a smali code string disassembled from an instruction
+        :return: a BytecodeObject
+        """
+        smali = smali.split("//")[0].strip()
```

Comment:
Using the split function may break the smali. For example, `smali.split("//")[0]` will return 'const v3, "string with' if `smali` is 'const v3, "string with //"'. We could change it to `smali = smali.rsplit("//", maxsplit=1)[0].strip()`.